streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java

```diff
@@ -316,59 +316,101 @@ public void shouldLogAndMeterWhenSkippingExpiredWindowByGrace() {
         final String topic = "topic";
 
         final KStream<String, String> stream1 = builder.stream(topic, Consumed.with(Serdes.String(), Serdes.String()));
-        stream1.groupByKey(Serialized.with(Serdes.String(), Serdes.String()))
-            .windowedBy(TimeWindows.of(ofMillis(10)).advanceBy(ofMillis(5)).grace(ofMillis(90L)))
-            .aggregate(
-                () -> "",
-                MockAggregator.toStringInstance("+"),
-                Materialized.<String, String, WindowStore<Bytes, byte[]>>as("topic1-Canonicalized").withValueSerde(Serdes.String()).withCachingDisabled().withLoggingDisabled()
-            )
-            .toStream()
-            .map((key, value) -> new KeyValue<>(key.toString(), value))
-            .to("output");
+        stream1.groupByKey(Grouped.with(Serdes.String(), Serdes.String()))
+               .windowedBy(TimeWindows.of(ofMillis(10)).advanceBy(ofMillis(10)).grace(ofMillis(90L)))
+               .aggregate(
+                   () -> "",
+                   MockAggregator.toStringInstance("+"),
+                   Materialized.<String, String, WindowStore<Bytes, byte[]>>as("topic1-Canonicalized").withValueSerde(Serdes.String()).withCachingDisabled().withLoggingDisabled()
+               )
+               .toStream()
+               .map((key, value) -> new KeyValue<>(key.toString(), value))
+               .to("output");
 
         LogCaptureAppender.setClassLoggerToDebug(KStreamWindowAggregate.class);
         final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
         try (final TopologyTestDriver driver = new TopologyTestDriver(builder.build(), props, 0L)) {
-            driver.pipeInput(recordFactory.create(topic, "k", "100", 100L));
-            driver.pipeInput(recordFactory.create(topic, "k", "0", 0L));
-            driver.pipeInput(recordFactory.create(topic, "k", "1", 1L));
-            driver.pipeInput(recordFactory.create(topic, "k", "2", 2L));
-            driver.pipeInput(recordFactory.create(topic, "k", "3", 3L));
-            driver.pipeInput(recordFactory.create(topic, "k", "4", 4L));
-            driver.pipeInput(recordFactory.create(topic, "k", "5", 5L));
+            driver.pipeInput(recordFactory.create(topic, "k", "100", 200L));
+            driver.pipeInput(recordFactory.create(topic, "k", "0", 100L));
+            driver.pipeInput(recordFactory.create(topic, "k", "1", 101L));
+            driver.pipeInput(recordFactory.create(topic, "k", "2", 102L));
+            driver.pipeInput(recordFactory.create(topic, "k", "3", 103L));
```

Comment:
For the `shouldLogAndMeterWhenSkippingExpiredWindowByGrace` and the `shouldLogAndMeterWhenSkippingExpiredWindow`  only the first record is not dropped (I know this is due to setting the stream time) and all subsequent records are dropped.  
Do we want to put a record in the middle that isn't late and thus not dropped?

Reply:
Just to clarify, the intent here is to test the logging and metering, not the dropping behavior. Are you suggesting that we should verify that the record in the middle does not get dropped *and therefore* it doesn't get logged or contribute to the metrics?

Reply:
Yes that was the intent of my suggestion, but considering you're clarification I think the test is fine as is.